cmd/snap-update-ns/utils.go

```diff
@@ -87,6 +89,49 @@ func IsReadOnly(dirFd int, dirName string) (bool, error) {
 	return false, nil
 }
 
+// IsTrustedTmpfs returns true if a directory belongs to a tmpfs mounted by snapd.
+//
+// The function inspects the directory (represented as both an open file
+// descriptor and the absolute path) and a list of changes that were applied to
+// the mount namespace. A directory is trusted if it is a tmpfs that was
+// mounted by snap-confine or snapd-update-ns. Note that sub-directories of a
+// trusted tmpfs are not considered trusted by this function.
+func IsTrustedTmpfs(dirFd int, dirName string, changes []*Change) (bool, error) {
+	var fsData syscall.Statfs_t
+	if err := sysFstatfs(dirFd, &fsData); err != nil {
+		return false, fmt.Errorf("cannot fstatfs %q: %s", dirName, err)
+	}
+	// If something is not a tmpfs it cannot be the trusted tmpfs we are looking for.
+	if fsData.Type != TmpfsMagic {
+		return false, nil
+	}
+	// Any of the past changes that mounted a tmpfs exactly at the directory we
+	// are inspecting is considered as trusted. This is conservative because it
+	// doesn't trust sub-directories of a trusted tmpfs. This approach is
+	// sufficient for the intended use.
+	//
+	// The algorithm goes over all the changes and picks up mount and unmount
+	// actions keeping track of what happened. The set of constraints in
+	// snap-update-ns and snapd prevent from mounting over an existing mount
+	// point so we don't need to consider e.g. a bind mount shadowing an active
+	// tmpfs.
+	mounted := false
+	for _, change := range changes {
+		if change.Entry.Type == "tmpfs" && change.Entry.Dir == dirName {
+			mounted = change.Action == Mount
+		}
+	}
+	if mounted {
+		return true, nil
+	}
+	// As a special exception, assume that a tmpfs over /var/lib is trusted.
+	// This tmpfs is created by snap-confine as a "quirk" to support a
+	// particular behavior of LXD.  Once the quirk is migrated to a mount
+	// profile (or removed entirely if no longer necessary) the following
+	// code fragment can go away.
+	return dirName == "/var/lib", nil
```

Comment:
Can you preface this with 'TODO:', 'XXX:', 'FIXME:' or similar? Also can you add a comment in the quirk code for something like 'When this changes, please also remove the quirk exception code in isSnapdCreatedPrivateTmpfs()'.